dali/kernels/slice/slice_flip_normalize_permute_pad_gpu.h

```diff
@@ -79,48 +127,103 @@ class SliceFlipNormalizePermutePadGpu {
     for (int i = 0; i < in.size(); i++) {
       const auto in_shape = in.tensor_shape(i);
-      auto processed_args = detail::ProcessArgs(args[i], in_shape);
-      if (i == 0) {
-        normalization_dim = processed_args.normalization_dim;
-      } else {
-        DALI_ENFORCE(normalization_dim == processed_args.normalization_dim);
-      }
+      auto &processed_args = processed_args_[i];
       auto &sample_desc = sample_descs_cpu[i];
       sample_desc.in_strides = processed_args.in_strides;
-      sample_desc.out_strides = processed_args.out_strides;
+      for (int d = 0; d < Dims; d++) sample_desc.out_strides[d] = processed_args.out_strides[d];
       sample_desc.out_shape = processed_args.out_shape;
-      sample_desc.padded_out_shape = processed_args.padded_out_shape;
-      sample_desc.padding_val = processed_args.padding_val;
+      sample_desc.in_shape = processed_args.in_shape;
+      sample_desc.anchor = processed_args.anchor;
+      sample_desc.norm_add = norm_add_gpu + i * norm_args_size_;
+      sample_desc.norm_mul = norm_mul_gpu + i * norm_args_size_;
+      sample_desc.fill_values = fill_values_gpu + i * norm_args_size_;
+      sample_desc.channel_dim = processed_args.channel_dim;
       sample_desc.in = in.tensor_data(i) + processed_args.input_offset;
       sample_desc.out = out.tensor_data(i);
-      sample_sizes[i] = volume(processed_args.padded_out_shape);
+      sample_desc.need_pad =
+          NeedPad(Dims, processed_args.anchor.data(), processed_args.in_shape.data(),
+                  processed_args.out_shape.data());
+      need_pad |= sample_desc.need_pad;
+      sample_desc.need_flip = false;
+      for (int d = 0; d < Dims; d++)
+        sample_desc.need_flip |= processed_args.in_strides[d] < 0;
+      need_flip |= sample_desc.need_flip;
+
+      // We fuse the last dimension with the previous IF:
+      // 1. There are at least 2 dimensions
+      // 2. Last dimension is not sliced/padded/permuted
+      // 3. Last dimension is not the channel dimension
+      // 4. The last two dimensions are either both flipped or not flipped
+      int last_dim = Dims - 1;
+      while (last_dim > 0 &&
+             sample_desc.out_strides[last_dim] == sample_desc.in_strides[last_dim] &&
+             sample_desc.anchor[last_dim] == 0 &&
+             sample_desc.out_shape[last_dim] == sample_desc.in_shape[last_dim] &&
+             sample_desc.channel_dim != last_dim &&
+             (sample_desc.in_strides[last_dim] < 0) == (sample_desc.in_strides[last_dim - 1] < 0)) {
+        last_dim--;
+      }
+
+      if (last_dim < Dims - 1) {
+        auto stride = sample_desc.in_strides[last_dim];  // same as out_strides[last_dim]
+        sample_desc.anchor[last_dim]    *= stride;
+        sample_desc.in_shape[last_dim]  *= stride;
+        sample_desc.out_shape[last_dim] *= stride;
+      }
+      sample_desc.effective_ndim = last_dim + 1;
+
+      sample_sizes[i] = volume(processed_args.out_shape);
```

Comment:
`sample_size` vector is not really necessary - it's elements are written once and then read once in the same order - you could calculate the volume directly
```c++
size_t remaining = volume(processed_args_[i].out_shape);
```
(GitHub won't let me put comment on that line :()